### ISSUE TITLE
rebuilding while offline

### ISSUE THREAD

**Issue by schmittlauch** [rycee/home-manager#430]
When using the default configuration where `programs.home-manager.path` points to a remote URL it is not possible to rebuild one's home configuration with `home-manager switch` when not connected to the internet.
I deliberately configured my system config to [be rebuildable even when offline](https://nixos.wiki/wiki/FAQ#How_to_keep_build-time_dependencies_around_.2F_be_able_to_rebuild_while_being_offline.3F ), so I ask for either an option to disable updating remote paths or for an autodetection of being offline/ unable to pull in the paths. (GitHub could be down, too)

This is especially important because even adding a line to one's vimrc using home-manager requires a rebuild and should not depend on internet connectivity.

**Comment by rycee** [comment]
Thanks for the report. Yeah, for this case I would suggest cloning the Home Manager repository and changing `programs.home-manager.path` to point to the absolute path of your clone. Something like

```nix
programs.home-manager.path = "$HOME/devel/home-manager";
```

The lack of control over remote imports is up to the Nix tools and cannot be controlled by Home Manager. Then update Home Manager using `git pull`.

**Comment by schmittlauch** [comment]
As a compromise it'd be nice to manage the home-manager sources via nix-channels as well. Updating the local git repo is an additional effort, but I update the nix-channels anyways.
Is that possible?

**Comment by rycee** [comment]
In principle it should straight forward to distribute Home Manager through a channel. But it would require setting up build infrastructure and hosting for the channel directory. Perhaps nixos.org would be willing to lend their hydra and CDN for the purpose?

In any case, I'm unlikely to start looking into this anytime soon due to work and other priorities but would be willing to assist anybody wanting to look into it.

**Comment by dtzWill** [comment]
On Tue, 23 Oct 2018 13:40:52 -0700, Robert Helgesson <notifications@github.com> wrote:
> In principle it should straight forward to distribute Home Manager through a channel. But it would require setting up build infrastructure and hosting for the channel directory. Perhaps nixos.org would be willing to lend their hydra and CDN for the purpose?
> 
> In any case, I'm unlikely to start looking into this anytime soon due to work and other priorities but would be willing to assist anybody wanting to look into it.

I think one of us needs to teach 'nix-channel' how to use a git repo
directly, it's a very common use case and AFAIK is the form in which
virtually all channels are created from anyway.

Or maybe just a bit of tooling to bridge the gap, dunno.

I know you can point it at the tarballs but that's a smidge inefficient
(full dl for every commit, also gz vs xz, etc.) but I want to say that it
also breaks something but I don't remember what-- only that I thought
it was great until I ran into this problem and now just have a bunch
of local git clones instead since apparently that's "better" :(.

Oh-- I'm not sure this is the only limitation, but one problem
is that nixpkgs acquired this way won't know what revision it is
which is annoying if not problematic.

**Comment by Mic92** [comment]
You could do the following to use the nix-channel command:

```console
$ nix-channel --add https://github.com/rycee/home-manager/archive/master.tar.gz home-manager
$ nix-channel --update
$ ls -la .nix-defexpr/channels/home-manager
```

```
programs.home-manager.path = "$HOME/.nix-defexpr/channels/home-manager";
```

**Comment by rycee** [comment]
@Mic92 That's nice! My understanding of the [channel format](https://nixos.org/nix/manual/#channel-format) was that it needed some special setup with a `binary-cache-url` file. If this works then I definitely would like to add it to the documentation. I will try it out as soon as possible. Thanks a lot!

**Comment by Mic92** [comment]
No the binary cache is orthogonal to the channel, which is literal just a tarball with a default.nix.

**Comment by rycee** [comment]
@Mic92 Created #475 to switch the installation instructions to recommend nix-channel instead. No need to set `programs.home-manager.path` if using home-manager as the channel name.

I've actually intended to support nix-channel for a long time but never got around to it because I assumed it would need some work on the CI side to create a correctly formatted directory: https://nixos.org/nix/manual/#channel-format

**Comment by rycee** [comment]
I merged #475 so I'll close this ticket. From now on the recommended way to set up Home Manager is using nix-channel and this will allow the Home Manager expression to be reliably available even while offline.
